### PR TITLE
[FIX] l10n_es_edi_tbai: invalid syntax error for expression

### PR DESCRIPTION
**Pull request by niyasraphy**
before this commit, if the user is running odoo with python 3.7, and when the function is getting executed, invalid syntax error for expression is shown.

the := walrus operator, was introduced in Python 3.8 and not available in python 3.7.

after this commit, the traceback wont be shown
and written expression will be valid for 3.7 and
above.

issue:  https://github.com/odoo/odoo/issues/122785

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by niyasraphy** [comment]
hi @odoostan ,
if you have time, can you review this, seems related to:  https://github.com/odoo/odoo/commit/5516ca7f77f14aaedf6565e8e8bd53789b0aafd1

thanks in advance

**Comment by MihranThalhath** [comment]
Since python 3.7 EOL is in 1 month, this might not be needed

**Comment by SlavekB** [comment]
> Since python 3.7 EOL is in 1 month, this might not be needed

It is use with Debian 10.x (Buster), which will have support until the June 30th 2024. It seems that for conservative users, it can be quite useful.

**Comment by odoostan** [comment]
Hi @niyasraphy, thanks for your PR.
Truth is, I only used the walrus operator because it was recently highlighted in a talk internally. Now that I look around the 16.0 codebase I see it's not used at all.
@jco-odoo what do you think ? Looking at the requirements it does seem like we should support python 3.7.

File: addons/l10n_es_edi_tbai/models/account_edi_format.py [hunk 1]
```diff
@@ -314,15 +314,20 @@ def _l10n_es_tbai_get_invoice_values(self, invoice, cancel):
 
         # Lines (detalle)
         refund_sign = (1 if values['is_refund'] else -1)
-        values['invoice_lines'] = [{
-            'line': line,
-            'discount': (discount := (line.balance / (1 - line.discount / 100) * line.discount / 100)) * refund_sign,
-            'unit_price': (line.balance + discount) / line.quantity * refund_sign,
-            'total': line.price_total * abs(line.balance / line.amount_currency if line.amount_currency != 0 else 1) * -refund_sign if not any(
-                [t.l10n_es_type == 'sujeto_isp' for t in line.tax_ids]) else abs(line.balance) * -refund_sign * (-1 if line.price_total < 0 else 1),
-            'description': regex_sub(r'[^0-9a-zA-Z ]', '', line.name)[:250],  # only keep characters allowed in description
```
Comment (odoostan) [inline addons/l10n_es_edi_tbai/models/account_edi_format.py:323]: Maybe it's not the most useful comment, but i'd rather improve it than remove it, e.g.
> description only allows up to 250 alphanumeric characters.

File: addons/l10n_es_edi_tbai/models/account_edi_format.py [hunk 2]
```diff
@@ -314,15 +314,20 @@ def _l10n_es_tbai_get_invoice_values(self, invoice, cancel):
...
-
+        invoice_lines = []
+        for line in invoice.invoice_line_ids.filtered(lambda line: line.display_type not in ('line_section', 'line_note')):
+            discount = line.balance / (1 - line.discount / 100) * line.discount / 100
+            invoice_lines.append({
+                'line': line,
+                'discount': discount * refund_sign,
+                'unit_price': (line.balance + discount) / line.quantity * refund_sign,
+                'total': line.price_total * abs(
+                line.balance / line.amount_currency if line.amount_currency != 0 else 1) * -refund_sign if not any(
```
[the remark this answers is not in the document]
Comment (niyasraphy) [reply addons/l10n_es_edi_tbai/models/account_edi_format.py]: @jco-odoo updated the pr.
checking the sujeto_isp was part of original commit: https://github.com/odoo/odoo/commit/5516ca7f77f14aaedf6565e8e8bd53789b0aafd1, may be @odoostan can provide some insights about it.

thanks

File: addons/l10n_es_edi_tbai/models/account_edi_format.py [hunk 3]
```diff
@@ -314,15 +314,17 @@ def _l10n_es_tbai_get_invoice_values(self, invoice, cancel):
...
-        } for line in invoice.invoice_line_ids.filtered(lambda line: line.display_type not in ('line_section', 'line_note'))]
-
+        invoice_lines = []
+        for line in invoice.invoice_line_ids.filtered(lambda line: line.display_type not in ('line_section', 'line_note')):
+            discount = line.balance / (1 - line.discount / 100) * line.discount / 100
+            invoice_lines.append({
+                'line': line,
+                'discount': discount * refund_sign,
+                'unit_price': (line.balance + discount) / line.quantity * refund_sign,
+                'total': line.price_total * abs(line.balance / line.amount_currency if line.amount_currency != 0 else 1) * -refund_sign if not any([t.l10n_es_type == 'sujeto_isp' for t in line.tax_ids]) else abs(line.balance) * -refund_sign * (-1 if line.price_total < 0 else 1),
```
Comment (jco-odoo) [inline addons/l10n_es_edi_tbai/models/account_edi_format.py:324]: It is ok to have multiple lines.  It is just that we need to see it is different than the other lines...

Comment (niyasraphy) [reply addons/l10n_es_edi_tbai/models/account_edi_format.py:324]: will it be fine, if you move it outside as follows:

        for line in invoice.invoice_line_ids.filtered(lambda line: line.display_type not in ('line_section', 'line_note')):
            discount = line.balance / (1 - line.discount / 100) * line.discount / 100
            if not any([t.l10n_es_type == 'sujeto_isp' for t in line.tax_ids]):
                total = line.price_total * abs(line.balance / line.amount_currency if line.amount_currency != 0 else 1) * -refund_sign
            else:
                total = abs(line.balance) * -refund_sign * (-1 if line.price_total < 0 else 1)
            invoice_lines.append({
                'line': line,
                'discount': discount * refund_sign,
                'unit_price': (line.balance + discount) / line.quantity * refund_sign,
                'total': total,
                'description': regex_sub(r'[^0-9a-zA-Z ]', '', line.name)[:250]
            })

Comment (niyasraphy) [reply addons/l10n_es_edi_tbai/models/account_edi_format.py]: updated

Comment (odoostan) [reply addons/l10n_es_edi_tbai/models/account_edi_format.py]: LGTM :+1:

File: addons/l10n_es_edi_tbai/models/account_edi_format.py (hunk 1 above)
Comment (odoostan) [reply addons/l10n_es_edi_tbai/models/account_edi_format.py:323]: I mean I wrote the suggestion and everything but it's not like a necessary comment. Whatever.

**Comment by jco-odoo** [comment]
@robodoo r+

**Comment by jco-odoo** [comment]
Thanks @niyasraphy

**Pull request closed without merging**